src/appleseed/renderer/modeling/postprocessingstage/effect/resamplex2applier.cpp

```diff
@@ -0,0 +1,142 @@
+//
+// This software is released under the MIT license.
+//
+// Copyright (c) 2020 Tiago Chaves, The appleseedhq Organization
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+// Interface header.
+#include "resamplex2applier.h"
+
+// appleseed.foundation headers.
+#include "foundation/image/canvasproperties.h"
+#include "foundation/image/color.h"
+#include "foundation/image/image.h"
+#include "foundation/math/scalar.h"
+
+// Standard headers.
+#include <cmath>
+
+using namespace foundation;
+
+namespace renderer
+{
+
+//
+// ResampleX2Applier class implementation.
+//
+
+ResampleX2Applier::ResampleX2Applier(
+    const Image&            src_image,
+    const SamplingMode      mode)
+  : m_mode(mode)
+  , m_src_width(src_image.properties().m_canvas_width)
+  , m_src_height(src_image.properties().m_canvas_height)
+  , m_src_image(src_image)
+{
+}
+
+void ResampleX2Applier::release()
+{
+    delete this;
+}
+
+void ResampleX2Applier::apply(
+    Image&              image,
+    const std::size_t   tile_x,
+    const std::size_t   tile_y) const
+{
+    assert(tile_x < image.properties().m_tile_count_x);
+    assert(tile_y < image.properties().m_tile_count_y);
+
+    Tile& tile = image.tile(tile_x, tile_y);
+    const std::size_t tile_width = tile.get_width();
+    const std::size_t tile_height = tile.get_height();
+    const Vector2u tile_offset(
+        tile_x * image.properties().m_tile_width,
+        tile_y * image.properties().m_tile_height);
+
+    const std::size_t dst_width = image.properties().m_canvas_width;
+    const std::size_t dst_height = image.properties().m_canvas_height;
+
+    const Vector2f scaling_factor(
+        static_cast<float>(m_src_width - 1) / (dst_width - 1),
+        static_cast<float>(m_src_height - 1) / (dst_height - 1));
+
+    assert(
+        m_mode == SamplingMode::DOUBLE
+        ? dst_width / 2 == m_src_width && dst_height / 2 == m_src_height    // scale x2 with bilinear filtering
```

Comment:
How can you ensure this test is robust regarding the fact you're using integers ? A more robust way to test it is `m_src_width * 2 == dst_width`.

Reply:
I'm using it this way to account for integer truncation, e.g.:
If the original image is 101x101, the first downscaling will result in a 50x50 image. Thus, we'll later have to upsample from 50x50 to 101x101, so `101 / 2 == 50` works but `50 * 2 == 101` wouldn't.

Reply:
Maybe you can explain this directly in the code ?